source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1119,43 +1172,218 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(ClusterManagerCluster& cm_
+  }
+}
+
+ClusterManagerImpl::ClusterInitializationObject::ClusterInitializationObject(
+    const absl::flat_hash_map<int, ThreadLocalClusterUpdateParams::PerPriority>& per_priority_state,
+    const ThreadLocalClusterUpdateParams& update_params, ClusterInfoConstSharedPtr cluster_info,
+    LoadBalancerFactorySharedPtr load_balancer_factory, HostMapConstSharedPtr map)
+    : per_priority_state_(per_priority_state), cluster_info_(std::move(cluster_info)),
+      load_balancer_factory_(load_balancer_factory), cross_priority_host_map_(map) {
+
+  ASSERT(cluster_info_->type() == envoy::config::cluster::v3::Cluster::EDS,
+         fmt::format("Using merge constructor on possible non-mergable cluster of type {}",
+                     cluster_info_->type()));
+  // Because EDS Clusters receive the entire ClusterLoadAssignment but only
+  // provides the delta we must process the hosts_added and hosts_removed and
+  // not simply overwrite with hosts added.
+  for (const auto& update : update_params.per_priority_update_params_) {
+    auto it = per_priority_state_.find(update.priority_);
+    if (it != per_priority_state_.end()) {
+      auto& priority_state = it->second;
+      // Merge the two per_priorities.
+      priority_state.update_hosts_params_ = update.update_hosts_params_;
+      priority_state.locality_weights_ = update.locality_weights_;
+      priority_state.weighted_priority_health_ = update.weighted_priority_health_;
+      priority_state.overprovisioning_factor_ = update.overprovisioning_factor_;
+
+      // Merge the hosts vectors to just have hosts added.
+      // Assumes that the old host_added_ is exclusive to new hosts_added_ and
+      // new hosts_removed_ only refers to the old hosts_added_.
+      ASSERT(priority_state.hosts_removed_.empty(),
+             "Cluster Initialization Object should apply hosts "
+             "removed updates to hosts_added vector!");
+
+      // TODO(kbaichoo): replace with a more efficient algorithm
+      if (!update.hosts_removed_.empty()) {
+        // Remove all hosts to be removed from the old host_added.
+        auto& host_added = priority_state.hosts_added_;
+        auto removed_section = std::remove_if(
+            host_added.begin(), host_added.end(),
+            [hosts_removed = std::cref(update.hosts_removed_)](const HostSharedPtr& ptr) {
+              return std::find(hosts_removed.get().begin(), hosts_removed.get().end(), ptr) !=
+                     hosts_removed.get().end();
+            });
+        priority_state.hosts_added_.erase(removed_section, priority_state.hosts_added_.end());
+      }
+
+      // Add updated host_added.
+      priority_state.hosts_added_.reserve(priority_state.hosts_added_.size() +
+                                          update.hosts_added_.size());
+      std::copy(update.hosts_added_.begin(), update.hosts_added_.end(),
+                std::back_inserter(priority_state.hosts_added_));
```

Comment:
I see how this works for newly updated hosts. Will this function also receive updates to hosts that already exist in the priority_state?